Visual Studio 2013 static library build as part of a bigger solution

# Issue description
Trying to use libzmq as part of a bigger solution fails for static library builds.
Though ZMQ_STATIC is set for all projects, the linker complains with:
4>czmq-static.lib(zframe.obj) : error LNK2019: unresolved external symbol __imp_zmq_msg_init referenced in function zframe_new
4>czmq-static.lib(zsock.obj) : error LNK2001: unresolved external symbol __imp_zmq_msg_init
....

The problem seems to be the build optimization using a cmake "object" library.
The "object" library also has to have ZMQ_STATIC set so that zmq.h is correctly preprocessed.


# Environment
Visual Studio 2013
Cmake 3.9 

* libzmq version: 1dd42fe
* OS: Windows 10

# Minimal test code / Steps to reproduce the issue

1.  cmake  -DBUILD_SHARED=OFF -DBUILD_STATIC=ON


# What's the actual result? (include assertion message & call stack if applicable)
4>czmq-static.lib(zframe.obj) : error LNK2019: unresolved external symbol __imp_zmq_msg_init referenced in function zframe_new
4>czmq-static.lib(zsock.obj) : error LNK2001: unresolved external symbol __imp_zmq_msg_init



# What's the expected result?
No linker errors :)


We are currently trying to fix this, and will provide a patch if possible.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -746,16 +746,12 @@ if (BUILD_STATIC)
   list(APPEND target_outputs "libzmq-static")
 endif()
 
-# avoid building everything twice for shared + static
-add_library (objects OBJECT ${sources})
-set_property(TARGET objects PROPERTY POSITION_INDEPENDENT_CODE ON)
-
 if (MSVC)
   # Suppress linker warnings caused by #ifdef omission
   # of file content.
   set( CMAKE_STATIC_LINKER_FLAGS /ignore:4221 )
   if (BUILD_SHARED)
-    add_library (libzmq SHARED $<TARGET_OBJECTS:objects> ${public_headers} ${html-docs} ${readme-docs} ${CMAKE_CURRENT_BINARY_DIR}/NSIS.template.in ${CMAKE_CURRENT_BINARY_DIR}/version.rc)
+    add_library (libzmq SHARED ${sources} ${public_headers} ${html-docs} ${readme-docs} ${CMAKE_CURRENT_BINARY_DIR}/NSIS.template.in ${CMAKE_CURRENT_BINARY_DIR}/version.rc)
     target_link_libraries (libzmq ${OPTIONAL_LIBRARIES})
     set_target_properties (libzmq PROPERTIES
       PUBLIC_HEADER "${public_headers}"
@@ -768,7 +764,7 @@ if (MSVC)
   endif()
   
   if (BUILD_STATIC)
-    add_library (libzmq-static STATIC $<TARGET_OBJECTS:objects> ${CMAKE_CURRENT_BINARY_DIR}/version.rc)
+    add_library (libzmq-static STATIC ${sources} ${CMAKE_CURRENT_BINARY_DIR}/version.rc)
     set_target_properties (libzmq-static PROPERTIES
       PUBLIC_HEADER "${public_headers}"
       RELEASE_POSTFIX "${MSVC_TOOLSET}-mt-s-${ZMQ_VERSION_MAJOR}_${ZMQ_VERSION_MINOR}_${ZMQ_VERSION_PATCH}"
@@ -778,6 +774,11 @@ if (MSVC)
       OUTPUT_NAME "libzmq")
   endif()
 else ()
+  # avoid building everything twice for shared + static
+  # only on *nix, as Windows needs different preprocessor defines in static builds
+  add_library (objects OBJECT ${sources})
+  set_property(TARGET objects PROPERTY POSITION_INDEPENDENT_CODE ON)
+
   if (BUILD_SHARED)
     add_library (libzmq SHARED $<TARGET_OBJECTS:objects> ${public_headers} ${html-docs} ${readme-docs} ${zmq-pkgconfig} ${CMAKE_CURRENT_BINARY_DIR}/version.rc)
     target_link_libraries (libzmq ${OPTIONAL_LIBRARIES})
@@ -812,9 +813,16 @@ else ()
       OUTPUT_NAME "zmq"
       PREFIX "lib")
   endif()
+
+  target_include_directories (objects
+    PUBLIC
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+      $<INSTALL_INTERFACE:include>
+  )
 endif ()
 
-foreach (target ${target_outputs} objects)
+foreach (target ${target_outputs})
   target_include_directories (${target}
     PUBLIC
       $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>

